src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+               .ToArray();
+            
+            var possibleAddresses =
+                (await Task.WhenAll(multiAddresses.Select(a => a.ResolveAsync(cancel))).ConfigureAwait(false))
+               .SelectMany(a => a)
+               .Where(a => !blackList.Contains(a.WithoutPeerId()))
+               .Select(a => a.WithPeerId(remote.Id))
+               .Distinct()
+               .ToArray();
+               
+            if (possibleAddresses.Length == 0)
+            {
+                throw new Exception($"{remote} has no known or reachable address.");
+            }
+
+            // Try the various addresses in parallel.  The first one to complete wins.
+            PeerConnection connection = null;
+            try
+            {
+                using (var timeout = new CancellationTokenSource(TransportConnectionTimeout))
+                {
+                    using (var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, cancel))
+                    {
+                        var attempts = possibleAddresses
+                           .Select(a => DialAsync(remote, a, cts.Token));
+                        connection = await TaskHelper.WhenAnyResultAsync(attempts, cts.Token).ConfigureAwait(false);
+                        cts.Cancel(); // stop other dialing tasks.
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                var attemped = string.Join(", ", possibleAddresses.Select(a => a.ToString()));
+                _log.Trace($"Cannot dial {attemped}");
+                throw new Exception($"Cannot dial {remote}.", e);
+            }
+
+            // Do the connection handshake.
+            try
+            {
+                MountProtocols(connection);
+                IEncryptionProtocol[] security;
+                lock (_protocols)
+                {
+                    security = _protocols.OfType<IEncryptionProtocol>().ToArray();
+                }
+
+                await connection.InitiateAsync(security, cancel).ConfigureAwait(false);
+                await connection.MuxerEstablished.Task.ConfigureAwait(false);
+                Identify1 identify;
+                lock (_protocols)
+                {
+                    identify = _protocols.OfType<Identify1>().First();
+                }
+
+                await identify.GetRemotePeerAsync(connection, cancel).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                connection.Dispose();
+                throw;
+            }
+
+            var actual = Manager.Add(connection);
+            if (actual == connection)
+            {
+                ConnectionEstablished?.Invoke(this, connection);
+            }
+
+            return actual;
+        }
+
+        private async Task<PeerConnection> DialAsync(Peer remote, MultiAddress addr, CancellationToken cancel)
+        {
+            // TODO: HACK: Currenty only the ipfs/p2p is supported.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)